File: 1460-number-of-substrings-containing-all-three-characters/solution.java
class Solution {
    public int numberOfSubstrings(String s) {
        int n = s.length();
        int i=0;
        int j=0;
        int count = 0;
        int[] freq = new int[3];

        while(j < n) {
            char ch = s.charAt(j);
            freq[ch-'a']++;

            while(freq['a'-'a'] >= 1 && freq['b'-'a'] >= 1 && freq['c'-'a'] >= 1) {
                count += n-j;
                freq[s.charAt(i)-'a']--;
                i++;
            }
            j++;
        }
        return count;
    }
}

File: 1460-number-of-substrings-containing-all-three-characters/solution.swift
class Solution {
    func numberOfSubstrings(_ s: String) -> Int {
        var n = s.count
        var i=0
        var j=0
        var ans = 0
        var dict = [Character: Int]()
        var chars = Array(s)

        while j < n {
            dict[chars[j], default: 0] += 1

            while dict["a", default: 0] >= 1 && 
                dict["b", default: 0] >= 1 && 
                dict["c", default: 0] >= 1 {
                // How many substrings START at i?
                ans += n-j
                // shrink
                dict[chars[i]]! -= 1
                i += 1
            }
            j += 1
        }
        return ans
    }
}
